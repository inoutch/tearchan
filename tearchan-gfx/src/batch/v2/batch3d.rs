use crate::batch::v2::buffer::BatchBuffer;
use crate::batch::v2::context::BatchContext;
use crate::batch::v2::provider::BatchProvider;
use crate::batch::v2::{Batch, BatchEvent};
use crate::v2::buffer::Buffer;
use nalgebra_glm::{Vec2, Vec3, Vec4};
use std::ops::{Deref, DerefMut};
use wgpu::RenderPass;

pub const BATCH3D_ATTRIBUTE_POSITION: u32 = 0;
pub const BATCH3D_ATTRIBUTE_TEXCOORD: u32 = 1;
pub const BATCH3D_ATTRIBUTE_COLOR: u32 = 2;
pub const BATCH3D_ATTRIBUTE_NORMAL: u32 = 3;

pub struct Batch3D {
    batch: Batch<Batch3DProvider>,
}

impl Deref for Batch3D {
    type Target = Batch<Batch3DProvider>;

    fn deref(&self) -> &Self::Target {
        &self.batch
    }
}

impl DerefMut for Batch3D {
    fn deref_mut(&mut self) -> &mut Self::Target {
        &mut self.batch
    }
}

impl Batch3D {
    pub fn new(device: &wgpu::Device) -> Self {
        let len: usize = 1024 * 100;
        let index_usage =
            wgpu::BufferUsages::INDEX | wgpu::BufferUsages::COPY_DST | wgpu::BufferUsages::COPY_SRC;
        let vertex_usage = wgpu::BufferUsages::VERTEX
            | wgpu::BufferUsages::COPY_DST
            | wgpu::BufferUsages::COPY_SRC;
        let provider = Batch3DProvider {
            index_buffer: BatchBuffer::new(Buffer::new(
                device,
                len,
                "IndexBuffer".to_string(),
                index_usage,
            )),
            position_buffer: BatchBuffer::new(Buffer::new(
                device,
                len,
                "PositionBuffer".to_string(),
                vertex_usage,
            )),
            texcoord_buffer: BatchBuffer::new(Buffer::new(
                device,
                len,
                "TexcoordBuffer".to_string(),
                vertex_usage,
            )),
            color_buffer: BatchBuffer::new(Buffer::new(
                device,
                len,
                "ColorBuffer".to_string(),
                vertex_usage,
            )),
            normal_buffer: BatchBuffer::new(Buffer::new(
                device,
                len,
                "NormalBuffer".to_string(),
                vertex_usage,
            )),
        };

        Batch3D {
            batch: Batch::new(provider, len, len),
        }
    }

    pub fn bind<'a>(&'a self, rpass: &mut RenderPass<'a>) {
        rpass.set_index_buffer(
            self.provider.index_buffer.buffer().slice(..),
            wgpu::IndexFormat::Uint32,
        );
        rpass.set_vertex_buffer(
            BATCH3D_ATTRIBUTE_POSITION as u32,
            self.provider.position_buffer.buffer().slice(..),
        );
        rpass.set_vertex_buffer(
            BATCH3D_ATTRIBUTE_TEXCOORD as u32,
            self.provider.texcoord_buffer.buffer().slice(..),
        );
        rpass.set_vertex_buffer(
            BATCH3D_ATTRIBUTE_COLOR as u32,
            self.provider.color_buffer.buffer().slice(..),
        );
        rpass.set_vertex_buffer(
            BATCH3D_ATTRIBUTE_NORMAL as u32,
            self.provider.normal_buffer.buffer().slice(..),
        );
    }
}

pub struct Batch3DProvider {
    index_buffer: BatchBuffer<Buffer<u32>, u32>,
    position_buffer: BatchBuffer<Buffer<Vec3>, Vec3>,
    texcoord_buffer: BatchBuffer<Buffer<Vec2>, Vec2>,
    color_buffer: BatchBuffer<Buffer<Vec4>, Vec4>,
    normal_buffer: BatchBuffer<Buffer<Vec3>, Vec3>,
}

impl<'a> BatchProvider<'a> for Batch3DProvider {
    type Context = BatchContext<'a>;

    fn run(&mut self, context: &mut Self::Context, event: BatchEvent) {
        match event {
            BatchEvent::WriteToIndexBuffer {
                pointer, object, ..
            } => {
                self.index_buffer.write(
                    context.writer(),
                    pointer,
                    &object.get_v1u32_indices().unwrap(),
                );
            }
            BatchEvent::WriteToVertexBuffer {
                pointer,
                object,
                attribute,
                ..
            } => {
                match attribute {
                    BATCH3D_ATTRIBUTE_POSITION => {
                        self.position_buffer.write(
                            context.writer(),
                            pointer,
                            &object.get_v3f32_vertices(attribute).unwrap(),
                        );
                    }
                    BATCH3D_ATTRIBUTE_TEXCOORD => {
                        self.texcoord_buffer.write(
                            context.writer(),
                            pointer,
                            &object.get_v2f32_vertices(attribute).unwrap(),
                        );
                    }
                    BATCH3D_ATTRIBUTE_COLOR => {
                        self.color_buffer.write(
                            context.writer(),
                            pointer,
                            &object.get_v4f32_vertices(attribute).unwrap(),
                        );
                    }
                    BATCH3D_ATTRIBUTE_NORMAL => {
                        self.normal_buffer.write(
                            context.writer(),
                            pointer,
                            &object.get_v3f32_vertices(attribute).unwrap(),
                        );
                    }
                    _ => {}
                };
            }
            BatchEvent::ClearToIndexBuffer { pointer } => {
                self.index_buffer.clear(context.writer(), pointer);
            }
            BatchEvent::ClearToVertexBuffer { pointer } => {
                self.position_buffer.clear(context.writer(), pointer);
                self.texcoord_buffer.clear(context.writer(), pointer);
                self.color_buffer.clear(context.writer(), pointer);
                self.normal_buffer.clear(context.writer(), pointer);
            }
            BatchEvent::ResizeIndexBuffer { len } => {
                self.index_buffer.resize(context.resizer(), len);
            }
            BatchEvent::ResizeVertexBuffer { len } => {
                self.position_buffer.resize(context.resizer(), len);
                self.texcoord_buffer.resize(context.resizer(), len);
                self.color_buffer.resize(context.resizer(), len);
                self.normal_buffer.resize(context.resizer(), len);
            }
            _ => {}
        }
    }
}
