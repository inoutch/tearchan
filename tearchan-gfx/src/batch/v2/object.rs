use crate::batch::types::{BatchAttributeIndex, BatchTypeArray, BatchTypeTransform};
use crate::batch::v2::buffer::BatchBufferPointer;
use nalgebra_glm::{TVec2, TVec3, TVec4};

pub struct BatchObject {
    pointer: BatchBufferPointer,
    data: Vec<BatchTypeArray>,
    transforms: Vec<BatchTypeTransform>,
    order: i32,
}

impl BatchObject {
    pub fn new(
        pointer: BatchBufferPointer,
        data: Vec<BatchTypeArray>,
        transforms: Vec<BatchTypeTransform>,
        order: i32,
    ) -> Self {
        BatchObject {
            pointer,
            data,
            transforms,
            order,
        }
    }

    pub fn order(&self) -> i32 {
        self.order
    }

    pub fn pointer(&self) -> BatchBufferPointer {
        self.pointer
    }

    pub fn data(&self) -> &Vec<BatchTypeArray> {
        &self.data
    }

    pub fn set_transform(&mut self, attribute: BatchAttributeIndex, transform: BatchTypeTransform) {
        self.transforms[attribute as usize] = transform;
    }

    pub fn set_data(&mut self, attribute: BatchAttributeIndex, data: BatchTypeArray) {
        self.data[attribute as usize] = data;
    }

    pub fn set_pointer(&mut self, pointer: BatchBufferPointer) {
        self.pointer = pointer;
    }

    pub fn get_transformed_data(
        &self,
        attribute: BatchAttributeIndex,
    ) -> Result<BatchTypeArray, BatchObjectError> {
        let data = self.data.get(attribute as usize).ok_or_else(|| {
            BatchObjectError::AttributeReferenceToDataOutOfRange {
                attribute,
                data_len: self.data.len(),
            }
        })?;
        if let Some(transform) = self.transforms.get(attribute as usize) {
            data.transform(transform)
                .ok_or(BatchObjectError::InvalidTransformType)
        } else {
            Ok(data.clone())
        }
    }

    pub fn get_v1f32_data(
        &self,
        attribute: BatchAttributeIndex,
    ) -> Result<Vec<f32>, BatchObjectError> {
        match self.get_transformed_data(attribute)? {
            BatchTypeArray::V1F32 { data } => Ok(data),
            _ => Err(BatchObjectError::InvalidArrayType),
        }
    }

    pub fn get_v1u32_data(
        &self,
        attribute: BatchAttributeIndex,
    ) -> Result<Vec<u32>, BatchObjectError> {
        match self.get_transformed_data(attribute)? {
            BatchTypeArray::V1U32 { data } => Ok(data),
            _ => Err(BatchObjectError::InvalidArrayType),
        }
    }

    pub fn get_v2f32_data(
        &self,
        attribute: BatchAttributeIndex,
    ) -> Result<Vec<TVec2<f32>>, BatchObjectError> {
        match self.get_transformed_data(attribute)? {
            BatchTypeArray::V2F32 { data } => Ok(data),
            _ => Err(BatchObjectError::InvalidArrayType),
        }
    }

    pub fn get_v2u32_data(
        &self,
        attribute: BatchAttributeIndex,
    ) -> Result<Vec<TVec2<u32>>, BatchObjectError> {
        match self.get_transformed_data(attribute)? {
            BatchTypeArray::V2U32 { data } => Ok(data),
            _ => Err(BatchObjectError::InvalidArrayType),
        }
    }

    pub fn get_v3f32_data(
        &self,
        attribute: BatchAttributeIndex,
    ) -> Result<Vec<TVec3<f32>>, BatchObjectError> {
        match self.get_transformed_data(attribute)? {
            BatchTypeArray::V3F32 { data } => Ok(data),
            _ => Err(BatchObjectError::InvalidArrayType),
        }
    }

    pub fn get_v3u32_data(
        &self,
        attribute: BatchAttributeIndex,
    ) -> Result<Vec<TVec3<u32>>, BatchObjectError> {
        match self.get_transformed_data(attribute)? {
            BatchTypeArray::V3U32 { data } => Ok(data),
            _ => Err(BatchObjectError::InvalidArrayType),
        }
    }

    pub fn get_v4f32_data(
        &self,
        attribute: BatchAttributeIndex,
    ) -> Result<Vec<TVec4<f32>>, BatchObjectError> {
        match self.get_transformed_data(attribute)? {
            BatchTypeArray::V4F32 { data } => Ok(data),
            _ => Err(BatchObjectError::InvalidArrayType),
        }
    }

    pub fn get_v4u32_data(
        &self,
        attribute: BatchAttributeIndex,
    ) -> Result<Vec<TVec4<u32>>, BatchObjectError> {
        match self.get_transformed_data(attribute)? {
            BatchTypeArray::V4U32 { data } => Ok(data),
            _ => Err(BatchObjectError::InvalidArrayType),
        }
    }
}

#[derive(Debug)]
pub enum BatchObjectError {
    AttributeReferenceToDataOutOfRange { attribute: u32, data_len: usize },
    InvalidTransformType,
    InvalidArrayType,
}
