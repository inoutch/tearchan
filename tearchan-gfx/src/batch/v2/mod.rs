use crate::batch::types::{BatchAttributeIndex, BatchTypeArray, BatchTypeTransform};
use crate::batch::v2::buffer::BatchBufferPointer;
use crate::batch::v2::object::BatchObject;
use crate::batch::v2::object_manager::{BatchObjectEvent, BatchObjectId, BatchObjectManager};
use crate::batch::v2::provider::BatchProvider;

pub mod batch2d;
pub mod batch3d;
pub mod batch_billboard;
pub mod batch_line;
pub mod buffer;
pub mod context;
pub mod object;
pub mod object_manager;
pub mod provider;

pub enum BatchEvent<'a> {
    Add {
        id: BatchObjectId,
    },
    Remove {
        id: BatchObjectId,
    },
    WriteToIndexBuffer {
        id: BatchObjectId,
        pointer: BatchBufferPointer,
        object: &'a BatchObject,
    },
    WriteToVertexBuffer {
        id: BatchObjectId,
        pointer: BatchBufferPointer,
        attribute: BatchAttributeIndex,
        object: &'a BatchObject,
    },
    ClearToIndexBuffer {
        pointer: BatchBufferPointer,
    },
    ClearToVertexBuffer {
        pointer: BatchBufferPointer,
    },
    ResizeIndexBuffer {
        len: usize,
    },
    ResizeVertexBuffer {
        len: usize,
    },
    Error,
}

pub struct Batch<T> {
    provider: T,
    manager: BatchObjectManager,
}

impl<T> Batch<T> {
    pub fn new(provider: T, index_len: usize, vertex_len: usize) -> Self {
        Batch {
            provider,
            manager: BatchObjectManager::new(index_len, vertex_len),
        }
    }

    pub fn provider(&self) -> &T {
        &self.provider
    }

    pub fn provider_mut(&mut self) -> &mut T {
        &mut self.provider
    }

    pub fn add(
        &mut self,
        indices: BatchTypeArray,
        vertices: Vec<BatchTypeArray>,
        order: Option<i32>,
    ) -> BatchObjectId {
        self.manager.add(indices, vertices, order)
    }

    pub fn remove(&mut self, id: BatchObjectId) {
        self.manager.remove(id);
    }

    pub fn replace_indices(&mut self, id: BatchObjectId, indices: BatchTypeArray) {
        self.manager.replace_indices(id, indices);
    }
    pub fn replace_vertices(&mut self, id: BatchObjectId, vertices: Vec<BatchTypeArray>) {
        self.manager.replace_vertices(id, vertices);
    }

    pub fn index_count(&self) -> usize {
        self.manager.index_allocator_len()
    }

    pub fn transform(
        &mut self,
        id: BatchObjectId,
        attribute: BatchAttributeIndex,
        transform: BatchTypeTransform,
    ) {
        self.manager.transform(id, attribute, transform);
    }
}

impl<TProvider> Batch<TProvider> {
    pub fn flush<'a>(&mut self, mut context: TProvider::Context)
    where
        TProvider: BatchProvider<'a>,
    {
        let provider = &mut self.provider;
        let manager = &mut self.manager;
        while let Some(event) = manager.pop_event() {
            match event {
                BatchObjectEvent::Add { id } => {
                    provider.run(&mut context, BatchEvent::Add { id });
                }
                BatchObjectEvent::Remove { id } => {
                    provider.run(&mut context, BatchEvent::Remove { id });
                }
                BatchObjectEvent::WriteToIndexBuffer { id } => {
                    let object = manager.get(id).unwrap();
                    provider.run(
                        &mut context,
                        BatchEvent::WriteToIndexBuffer {
                            id,
                            pointer: object.index_pointer(),
                            object,
                        },
                    );
                }
                BatchObjectEvent::WriteToVertexBuffer { id, attribute } => {
                    let object = manager.get(id).unwrap();
                    provider.run(
                        &mut context,
                        BatchEvent::WriteToVertexBuffer {
                            id,
                            pointer: object.vertex_pointer(),
                            attribute,
                            object,
                        },
                    );
                }
                BatchObjectEvent::ClearToIndexBuffer { pointer } => {
                    provider.run(&mut context, BatchEvent::ClearToIndexBuffer { pointer });
                }
                BatchObjectEvent::ClearToVertexBuffer { pointer } => {
                    provider.run(&mut context, BatchEvent::ClearToVertexBuffer { pointer });
                }
                BatchObjectEvent::ResizeIndexBuffer { len } => {
                    provider.run(&mut context, BatchEvent::ResizeIndexBuffer { len });
                }
                BatchObjectEvent::ResizeVertexBuffer { len } => {
                    provider.run(&mut context, BatchEvent::ResizeVertexBuffer { len });
                }
            }
        }
    }
}

#[cfg(test)]
mod test {
    use crate::batch::types::BatchTypeArray;
    use crate::batch::v2::buffer::BatchBuffer;
    use crate::batch::v2::provider::BatchProvider;
    use crate::batch::v2::{Batch, BatchEvent};
    use crate::v2::buffer::test::{TestBuffer, TestCopier, TestResizer, TestWriter};
    use nalgebra_glm::{vec2, vec3, Vec2, Vec3};
    use std::ops::{Deref, DerefMut};

    #[allow(dead_code)]
    struct TestBatchContext<'a> {
        resizer: &'a mut TestResizer,
        writer: &'a mut TestWriter,
        copier: &'a mut TestCopier,
    }

    struct TestBatch(Batch<TestBatchProvider>);

    impl Deref for TestBatch {
        type Target = Batch<TestBatchProvider>;

        fn deref(&self) -> &Self::Target {
            &self.0
        }
    }

    impl DerefMut for TestBatch {
        fn deref_mut(&mut self) -> &mut Self::Target {
            &mut self.0
        }
    }

    impl TestBatch {
        pub fn new(index_len: usize, vertex_len: usize) -> TestBatch {
            TestBatch(Batch::new(
                TestBatchProvider {
                    index_buffer: BatchBuffer::new(TestBuffer::new(vec![0; index_len])),
                    position_buffer: BatchBuffer::new(TestBuffer::new(
                        vec![(); vertex_len]
                            .iter()
                            .map(|_| vec3(0.0f32, 0.0f32, 0.0f32))
                            .collect(),
                    )),
                    texcoord_buffer: BatchBuffer::new(TestBuffer::new(
                        vec![(); vertex_len]
                            .iter()
                            .map(|_| vec2(0.0f32, 0.0f32))
                            .collect(),
                    )),
                },
                index_len,
                vertex_len,
            ))
        }
    }

    struct TestBatchProvider {
        index_buffer: BatchBuffer<TestBuffer<u32>, u32>,
        position_buffer: BatchBuffer<TestBuffer<Vec3>, Vec3>,
        texcoord_buffer: BatchBuffer<TestBuffer<Vec2>, Vec2>,
    }

    impl<'a> BatchProvider<'a> for TestBatchProvider {
        type Context = (&'a mut TestWriter, &'a mut TestCopier, &'a mut TestResizer);

        fn run(&mut self, context: &mut Self::Context, event: BatchEvent) {
            match event {
                BatchEvent::WriteToIndexBuffer {
                    pointer, object, ..
                } => {
                    self.index_buffer.write(
                        context.0,
                        pointer,
                        &object.get_v1u32_indices().unwrap(),
                    );
                }
                BatchEvent::WriteToVertexBuffer {
                    pointer,
                    attribute,
                    object,
                    ..
                } => match attribute {
                    0 => {
                        self.position_buffer.write(
                            context.0,
                            pointer,
                            &object.get_v3f32_vertices(attribute).unwrap(),
                        );
                    }
                    1 => self.texcoord_buffer.write(
                        context.0,
                        pointer,
                        &object.get_v2f32_vertices(attribute).unwrap(),
                    ),
                    _ => {}
                },
                BatchEvent::ClearToIndexBuffer { pointer } => {
                    self.index_buffer.clear(context.0, pointer);
                }
                BatchEvent::ClearToVertexBuffer { pointer } => {
                    self.position_buffer.clear(context.0, pointer);
                }
                BatchEvent::ResizeIndexBuffer { len } => {
                    self.index_buffer.resize(context.2, len);
                }
                BatchEvent::ResizeVertexBuffer { len } => {
                    self.position_buffer.resize(context.2, len);
                }
                _ => {}
            }
        }
    }

    #[test]
    fn test_add() {
        let mut resizer = TestResizer;
        let mut writer = TestWriter;
        let mut copier = TestCopier;

        let mut batch = TestBatch::new(10, 10);
        batch.add(
            BatchTypeArray::V1U32 {
                data: vec![0, 1, 2],
            },
            vec![BatchTypeArray::V3F32 {
                data: vec![
                    vec3(0.0f32, 0.0f32, 0.0f32),
                    vec3(1.0f32, 0.0f32, 0.0f32),
                    vec3(1.0f32, 1.0f32, 0.0f32),
                ],
            }],
            None,
        );

        batch.flush((&mut writer, &mut copier, &mut resizer));

        assert_eq!(
            batch.provider().index_buffer.buffer().data.borrow()[0..batch.index_count()],
            vec![0, 1, 2]
        );
        assert_eq!(
            batch.provider().position_buffer.buffer().data.borrow()[0..batch.index_count()],
            vec![
                vec3(0.0f32, 0.0f32, 0.0f32),
                vec3(1.0f32, 0.0f32, 0.0f32),
                vec3(1.0f32, 1.0f32, 0.0f32)
            ]
        );
    }
}
