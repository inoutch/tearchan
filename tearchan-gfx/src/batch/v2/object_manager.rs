use crate::batch::types::{BatchAttributeIndex, BatchTypeArray, BatchTypeTransform};
use crate::batch::v2::buffer::{
    BatchBufferAllocator, BatchBufferAllocatorEvent, BatchBufferPointer,
};
use crate::batch::v2::object::BatchObject;
use crate::batch::DEFAULT_ORDER;
use std::collections::{HashMap, HashSet, VecDeque};
use tearchan_util::id_manager::IdManager;

#[derive(Hash, Eq, PartialEq, Copy, Clone, Debug)]
pub struct BatchObjectId(u64);

impl BatchObjectId {
    fn next(&self) -> Self {
        BatchObjectId(self.0 + 1)
    }
}

#[derive(Debug)]
pub enum BatchObjectEvent {
    Add {
        id: BatchObjectId,
    },
    Remove {
        id: BatchObjectId,
    },
    Write {
        id: BatchObjectId,
        attribute: BatchAttributeIndex,
    },
    Clear {
        pointer: BatchBufferPointer,
    },
    Resize {
        len: usize,
    },
}

pub struct BatchObjectManager {
    id_manager: IdManager<BatchObjectId>,
    objects: HashMap<BatchObjectId, BatchObject>,
    object_ids_grouped_by_pointer: HashMap<BatchBufferPointer, BatchObjectId>,
    objects_will_be_rewritten: HashSet<(BatchObjectId, BatchAttributeIndex)>,
    allocator: BatchBufferAllocator,
    events: VecDeque<BatchObjectEvent>,
    len: usize,
}

impl BatchObjectManager {
    pub fn new(len: usize) -> Self {
        BatchObjectManager {
            id_manager: IdManager::new(BatchObjectId(0), |id| id.next()),
            objects: HashMap::new(),
            objects_will_be_rewritten: HashSet::new(),
            object_ids_grouped_by_pointer: HashMap::new(),
            allocator: BatchBufferAllocator::default(),
            events: VecDeque::new(),
            len,
        }
    }

    pub fn pop_event(&mut self) -> Option<BatchObjectEvent> {
        while let Some(event) = self.allocator.pop_event() {
            match event {
                BatchBufferAllocatorEvent::Write(pointer) => {
                    let object_id = self.object_ids_grouped_by_pointer.get(&pointer).unwrap();
                    let object = self.objects.get(object_id).unwrap();
                    for (i, _) in object.data().iter().enumerate() {
                        self.events.push_back(BatchObjectEvent::Write {
                            id: *object_id,
                            attribute: i as BatchAttributeIndex,
                        });
                    }
                }
                BatchBufferAllocatorEvent::Clear(pointer) => {
                    self.events.push_back(BatchObjectEvent::Clear { pointer });
                }
                BatchBufferAllocatorEvent::Reallocate { from, to } => {
                    let object_id = *self.object_ids_grouped_by_pointer.get(&from).unwrap();
                    let object = self.objects.get_mut(&object_id).unwrap();
                    self.object_ids_grouped_by_pointer.remove(&object.pointer());

                    object.set_pointer(to);
                    self.object_ids_grouped_by_pointer.insert(to, object_id);

                    for i in 0u32..object.data().len() as u32 {
                        self.objects_will_be_rewritten.insert((object_id, i));
                    }
                }
            }
        }
        let event = self.events.pop_front()?;
        match &event {
            BatchObjectEvent::Add { .. } => {}
            BatchObjectEvent::Remove { .. } => {}
            BatchObjectEvent::Write { id, attribute } => {
                self.objects_will_be_rewritten.remove(&(*id, *attribute));
            }
            BatchObjectEvent::Clear { .. } => {}
            BatchObjectEvent::Resize { .. } => {}
        }
        Some(event)
    }

    pub fn clear_events(&mut self) {
        self.objects_will_be_rewritten.clear();
    }

    #[inline]
    pub fn get(&self, id: BatchObjectId) -> Option<&BatchObject> {
        self.objects.get(&id)
    }

    pub fn add(
        &mut self,
        data: Vec<BatchTypeArray>,
        len: usize,
        order: Option<i32>,
    ) -> BatchObjectId {
        for datum in data.iter() {
            assert_eq!(datum.len(), len);
        }

        let id = self.id_manager.gen();
        let order = order.unwrap_or(DEFAULT_ORDER);
        let transforms = vec![BatchTypeTransform::None; data.len()];
        let pointer = self.allocator.allocate(len);
        let object = BatchObject::new(pointer, data, transforms, order);

        if self.allocator.len() > self.len {
            self.len = self.allocator.len() * 2;
            self.events
                .push_back(BatchObjectEvent::Resize { len: self.len });
        }

        for i in 0u32..object.data().len() as u32 {
            self.objects_will_be_rewritten.insert((id, i));
        }
        self.events.push_back(BatchObjectEvent::Add { id });
        self.objects.insert(id, object);
        self.object_ids_grouped_by_pointer.insert(pointer, id);
        id
    }

    pub fn remove(&mut self, id: BatchObjectId) -> Option<BatchObject> {
        self.events.push_back(BatchObjectEvent::Remove { id });
        let object = self.objects.remove(&id)?;
        self.allocator.free(object.pointer());
        self.object_ids_grouped_by_pointer.remove(&object.pointer());
        for attribute in 0..object.data().len() {
            self.objects_will_be_rewritten
                .remove(&(id, attribute as u32));
        }
        Some(object)
    }

    pub fn transform(
        &mut self,
        id: BatchObjectId,
        attribute: BatchAttributeIndex,
        transform: BatchTypeTransform,
    ) {
        let object = match self.objects.get_mut(&id) {
            Some(object) => object,
            None => return,
        };
        object.set_transform(attribute, transform);

        let key = (id, attribute);
        if self.objects_will_be_rewritten.get(&key).is_none() {
            self.events
                .push_back(BatchObjectEvent::Write { id, attribute });
            self.objects_will_be_rewritten.insert(key);
        }
    }

    pub fn rewrite(
        &mut self,
        id: BatchObjectId,
        attribute: BatchAttributeIndex,
        data: BatchTypeArray,
    ) {
        let object = match self.objects.get_mut(&id) {
            Some(object) => object,
            None => return,
        };
        let prev_data = match object.data().get(attribute as usize) {
            Some(data) => data,
            None => return,
        };
        assert_eq!(data.len(), prev_data.len());

        object.set_data(attribute, data);
        let key = (id, attribute);
        if self.objects_will_be_rewritten.get(&key).is_none() {
            self.events
                .push_back(BatchObjectEvent::Write { id, attribute });
            self.objects_will_be_rewritten.insert(key);
        }
    }

    pub fn replace(&mut self, id: BatchObjectId, len: usize, data: Vec<BatchTypeArray>) {
        for datum in data.iter() {
            assert_eq!(datum.len(), len);
        }

        let object = match self.objects.get_mut(&id) {
            Some(object) => object,
            None => return,
        };
        if !object.data().is_empty() && object.data().get(0).unwrap().len() == len {
            for (i, data) in data.into_iter().enumerate() {
                self.rewrite(id, i as u32, data);
            }
            return;
        }

        self.object_ids_grouped_by_pointer.remove(&object.pointer());

        let new_pointer = self.allocator.reallocate(object.pointer(), len);
        object.set_pointer(new_pointer);
        self.object_ids_grouped_by_pointer.insert(new_pointer, id);

        for i in 0u32..object.data().len() as u32 {
            self.objects_will_be_rewritten.insert((id, i));
        }
    }

    pub fn allocator_len(&self) -> usize {
        self.allocator.len()
    }

    pub fn allocator_is_empty(&self) -> bool {
        self.allocator.is_empty()
    }
}

#[cfg(test)]
mod test {
    use crate::batch::types::{BatchTypeArray, BatchTypeTransform};
    use crate::batch::v2::object_manager::{BatchObjectEvent, BatchObjectManager};
    use nalgebra_glm::{vec2, vec3, Mat2, Mat3};

    fn convert_events(manager: &mut BatchObjectManager) -> Vec<BatchObjectEvent> {
        let mut events = Vec::new();
        while let Some(event) = manager.pop_event() {
            events.push(event);
        }
        events
    }

    #[test]
    fn test_valid_len() {
        let mut manager = BatchObjectManager::new(100);
        manager.add(
            vec![
                BatchTypeArray::V1F32 { data: vec![0.0f32] },
                BatchTypeArray::V2F32 {
                    data: vec![vec2(0.0f32, 0.0f32)],
                },
            ],
            1,
            None,
        );
    }

    #[test]
    #[should_panic]
    fn test_invalid_len() {
        let mut manager = BatchObjectManager::new(100);
        manager.add(
            vec![
                BatchTypeArray::V1F32 {
                    data: vec![0.0f32, 0.0f32],
                },
                BatchTypeArray::V2F32 {
                    data: vec![vec2(0.0f32, 0.0f32)],
                },
            ],
            2,
            None,
        );
    }

    #[test]
    fn test_cleanup() {
        let mut manager = BatchObjectManager::new(100);
        let data = vec![
            BatchTypeArray::V1F32 { data: vec![0.0f32] },
            BatchTypeArray::V2F32 {
                data: vec![vec2(0.0f32, 0.0f32)],
            },
        ];
        let id0 = manager.add(data, 1, None);

        let data = vec![
            BatchTypeArray::V1F32 { data: vec![0.0f32] },
            BatchTypeArray::V2F32 {
                data: vec![vec2(0.0f32, 0.0f32)],
            },
        ];
        let id1 = manager.add(data, 1, None);
        manager.remove(id0);
        manager.remove(id1);

        assert_eq!(manager.objects.len(), 0);
        assert_eq!(manager.object_ids_grouped_by_pointer.len(), 0);
        assert_eq!(manager.objects_will_be_rewritten.len(), 0);
        assert_eq!(manager.allocator.len(), 0);
    }

    #[test]
    fn test_events() {
        let mut manager = BatchObjectManager::new(100);
        let id0 = manager.add(
            vec![
                BatchTypeArray::V1F32 { data: vec![1.0f32] },
                BatchTypeArray::V2F32 {
                    data: vec![vec2(1.0f32, 1.0f32)],
                },
            ],
            1,
            None,
        );
        let id1 = manager.add(
            vec![
                BatchTypeArray::V1F32 { data: vec![2.0f32] },
                BatchTypeArray::V2F32 {
                    data: vec![vec2(2.0f32, 2.0f32)],
                },
            ],
            1,
            None,
        );

        insta::assert_debug_snapshot!(convert_events(&mut manager));

        manager.transform(
            id0,
            1,
            BatchTypeTransform::Mat3F32 {
                m: Mat3::new(
                    1.0f32, 0.0f32, 0.0f32, 0.0f32, 1.0f32, 0.0f32, 0.0f32, 0.0f32, 1.0f32,
                ),
            },
        );
        manager.transform(
            id1,
            0,
            BatchTypeTransform::Mat2F32 {
                m: Mat2::new(1.0f32, 0.0f32, 0.0f32, 1.0f32),
            },
        );

        insta::assert_debug_snapshot!(convert_events(&mut manager));

        manager.remove(id0);

        insta::assert_debug_snapshot!(convert_events(&mut manager));

        manager.remove(id1);

        insta::assert_debug_snapshot!(convert_events(&mut manager));
    }

    #[test]
    fn test_resize_event() {
        let mut manager = BatchObjectManager::new(2);
        let _id0 = manager.add(
            vec![
                BatchTypeArray::V1U32 { data: vec![1, 1] },
                BatchTypeArray::V2F32 {
                    data: vec![vec2(1.0f32, 1.0f32), vec2(1.0f32, 1.0f32)],
                },
                BatchTypeArray::V3F32 {
                    data: vec![vec3(1.0f32, 1.0f32, 1.0f32), vec3(1.0f32, 1.0f32, 1.0f32)],
                },
            ],
            2,
            None,
        );

        insta::assert_debug_snapshot!(convert_events(&mut manager));

        let _id1 = manager.add(
            vec![
                BatchTypeArray::V1U32 {
                    data: vec![2, 2, 2],
                },
                BatchTypeArray::V2F32 {
                    data: vec![
                        vec2(2.0f32, 2.0f32),
                        vec2(2.0f32, 2.0f32),
                        vec2(2.0f32, 2.0f32),
                    ],
                },
                BatchTypeArray::V3F32 {
                    data: vec![
                        vec3(2.0f32, 2.0f32, 2.0f32),
                        vec3(2.0f32, 2.0f32, 2.0f32),
                        vec3(2.0f32, 2.0f32, 2.0f32),
                    ],
                },
            ],
            3,
            None,
        );
        insta::assert_debug_snapshot!(convert_events(&mut manager));
    }
}
