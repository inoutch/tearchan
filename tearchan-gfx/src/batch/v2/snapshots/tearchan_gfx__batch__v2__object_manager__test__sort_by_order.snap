---
source: tearchan-gfx/src/batch/v2/object_manager.rs
expression: convert_events(&mut manager)

---
[
    Add {
        id: BatchObjectId(
            0,
        ),
    },
    ResizeIndexBuffer {
        len: 4,
    },
    ResizeVertexBuffer {
        len: 8,
    },
    Add {
        id: BatchObjectId(
            1,
        ),
    },
    Add {
        id: BatchObjectId(
            2,
        ),
    },
    WriteToIndexBuffer {
        id: BatchObjectId(
            1,
        ),
    },
    WriteToIndexBuffer {
        id: BatchObjectId(
            2,
        ),
    },
    WriteToIndexBuffer {
        id: BatchObjectId(
            0,
        ),
    },
    WriteToVertexBuffer {
        id: BatchObjectId(
            0,
        ),
        attribute: 0,
    },
    WriteToVertexBuffer {
        id: BatchObjectId(
            0,
        ),
        attribute: 1,
    },
    WriteToVertexBuffer {
        id: BatchObjectId(
            0,
        ),
        attribute: 2,
    },
    WriteToVertexBuffer {
        id: BatchObjectId(
            1,
        ),
        attribute: 0,
    },
    WriteToVertexBuffer {
        id: BatchObjectId(
            1,
        ),
        attribute: 1,
    },
    WriteToVertexBuffer {
        id: BatchObjectId(
            1,
        ),
        attribute: 2,
    },
    WriteToVertexBuffer {
        id: BatchObjectId(
            2,
        ),
        attribute: 0,
    },
    WriteToVertexBuffer {
        id: BatchObjectId(
            2,
        ),
        attribute: 1,
    },
    WriteToVertexBuffer {
        id: BatchObjectId(
            2,
        ),
        attribute: 2,
    },
]
